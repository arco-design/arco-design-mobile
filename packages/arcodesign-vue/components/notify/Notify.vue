<script lang="ts" setup>
import { nextTick, onMounted, onUnmounted, ref, watch } from 'vue';
import { getPrefixCls } from '../context-provider';
import { NotifyEmits, NotifyProps } from './type';

const props = withDefaults(defineProps<NotifyProps>(), {
    type: 'info',
    transitionDuration: 300,
    duration: 3000,
});
const emit = defineEmits<NotifyEmits>();
const visible = defineModel<boolean>();
const prefix = getPrefixCls('notify');
const hasTrans = ref(Boolean(visible.value));
const innerHeight = ref(0);
const closeTimer = ref<number>();
const emitCloseTimer = ref<number>();
const innerDom = ref<HTMLDivElement>();

function updateLayout() {
    if (innerDom.value) {
        const { height } = innerDom.value.getBoundingClientRect();
        innerHeight.value = height;
    }
}

function startCloseTimer(duration: number) {
    if (duration) {
        clearTimeout(closeTimer.value);
        closeTimer.value = window.setTimeout(() => {
            visible.value = false;
        }, duration);
    }
}

function handleAfterVisible(duration: number) {
    nextTick(() => {
        updateLayout();
        startCloseTimer(duration);
    });
}

onMounted(() => {
    hasTrans.value = true;
    visible.value && handleAfterVisible(props.duration);
});

onUnmounted(() => {
    clearTimeout(emitCloseTimer.value);
    clearTimeout(closeTimer.value);
});

watch(visible, (newVisible) => {
    if (!newVisible) {
        emitCloseTimer.value = window.setTimeout(() => {
            emit('close');
        }, props.transitionDuration);
    }
});

watch([visible, () => props.duration], ([newVisible, newDuration], [oldVisible]) => {
    oldVisible && clearTimeout(closeTimer.value);
    newVisible && handleAfterVisible(newDuration);
});
</script>

<template>
    <Teleport :disabled="!props.container" :to="props.container" >
        <div
            :class="[prefix, `${prefix}-${props.type}`]"
            :style="{
                ...(hasTrans
                    ? { transitionDuration: `${props.transitionDuration}ms` }
                    : {}),
                height: visible ? `${innerHeight}px` : '0px',
            }"
        >
            <div
                v-if="$slots.default"
                :class="[
                    `${prefix}-content`,
                    visible
                        ? `${prefix}-content-transition-Y0`
                        : `${prefix}-content-transition-Y100`
                ]"
                :style="{
                    ...(hasTrans
                    ? { transitionDuration: `${props.transitionDuration}ms` }
                    : {}),
                }"
                ref="innerDom"
            >
                <slot></slot>
            </div>
        </div>
    </Teleport>
</template>
