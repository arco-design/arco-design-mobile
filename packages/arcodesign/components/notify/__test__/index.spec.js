import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import Notify from '..';

demoTest('notify');

mountTest(Notify, 'Notify');
