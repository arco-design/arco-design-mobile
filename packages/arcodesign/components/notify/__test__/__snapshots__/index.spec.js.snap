// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div
    class="arco-notify arco-notify-info "
    style="height: 0px;"
  >
    <div
      class="arco-notify-content arco-notify-content-transition-Y0"
    >
      Hello, world!
    </div>
  </div>
</div>
`;
